store/tikv/scan.go

```diff
@@ -94,11 +96,13 @@ func (s *Scanner) Next() error {
 				continue
 			}
 		}
-		if err := s.resolveCurrentLock(bo); err != nil {
+		resolved, err := s.resolveCurrentLock(bo)
+		if err != nil {
 			s.Close()
 			return errors.Trace(err)
 		}
-		if len(s.Value()) == 0 {
+
+		if resolved && len(s.Value()) == 0 {
```

Comment:
If `resolved` is false from https://github.com/pingcap/tidb/pull/7419/files#diff-d1ae066017b8533eea6b56eb55c4382cR125, the logic is different from the old code.

Reply:
It is different, and it is more appropriate and clear. The value returned from a seek cannot be nil if none errors happened.

Reply:
IMHO, when several conditions are combined, it's not more appropriate and clear.

`resolved == false` is ambiguous: is there no lock, so no need to resolve lock? 
or is there a lock, and resolved fail, so it's false ?

When `resolved && len(s.Value())` was written down, I have to consider four combination:

* resolve success && key not exist 
* resolve success && key exist
* resolve fail &&  key not exists
* resolve fail && key exist 

I think it's more appropriate to split the conditions, check lock first, then decide whether to resolve lock, and check key results afterwards.

Reply:
I love your ideas. It is actually not human-friendly when combines several conditions. I did that for keeping the same style as the old code.

Let me explain what I thought more in details, and then talk about how to improve it.

I said more `appropriate and clear` according to the old code, it is not an argument about `if conditions`. 

The source of evil here is the method `resolveCurrentLock `, it is completely side effect. It is invoked without any condition and checks GetError() inside, furthermore, it does not take any arguments and return any results, but it indeed changes the current item of the iterator. 

There is nothing relation between `len(s.Value())` and `resolveCurrentLock` we can tell from the original code, but it actually has. The `resolved` I added indicates the relation, that's why I said `more appropriate and clear`. I want to focus on the `KeyOnly` feature in this PR, so I did not change `resolveCurrentLock` too much, for achieving my goal and keeping the old code style.

Now I think I made it clear, let's talk about the solution. Maybe we have two options
1. Focus on the `KeyOnly` feature here and using the way I did. Maybe we can refactor `resolveCurrentLock` in another PR.
2. Refactor `resolveCurrentLock` and do the way as your purpose. All in this PR.

These two options are all OK for me, so I am looking for your suggestions.

Reply:
> it is completely side effect

Good catch!

>  I want to focus on the KeyOnly feature in this PR

Or we can reset unrelated changes

> 2. Refactor resolveCurrentLock and do the way as your purpose. All in this PR.

I think option 2 is acceptable, this refactor is just a tiny change. @shafreeck

Reply:
OK, I will take another commit.